File: examples/binary-search-c/bsearch.c

/*W logic In : int array,int,int,int -> prop */

int binary_search(int t[], int n, int v) 
/*@ array_length(t) = n and n >= 0 and sorted_array(t,0,n-1) */
{
  int l,u,p,m;
  l = 0; 
  u = n-1; 
  p = -1;
  while (l <= u) 
    /*@ invariant 
	  0 <= l and u <= array_length(t)-1 and -1 <= p <= array_length(t)-1
          and (p = -1 -> In(t,0,array_length(t)-1,v) -> In(t,l,u,v))
          and (p >= 0 -> t[p]=v)
        variant 2+u-l */ {
    m = (l + u) / 2;
    /*@ assert l <= m and m <= u */;
    if (t[m] < v)
      l = m + 1;
    else if (t[m] > v)
      u = m - 1;
    else
      return m;
  }
  return -1;
}
/*@ (0 <= result <= array_length(t)-1 and t[result]=v) or 
    (result = -1 and not In(t,1,array_length(t)-1,v)) */

File: examples/binary-search/bsearch.ml

external N : int
external v : int

external mean : int -> int -> int

external t : array (N+1) of int

external l,u,p,m : int ref

let binary_search =
  { sorted_array(t,1,N) }
  begin
    l := 1; u := N; p := 0;
    while !l <= !u do
      { invariant 1 <= l and u <= N and 0 <= p and p <= N
               and (p = 0 -> In(t,1,N) -> In(t,l,u))
               and (p > 0 -> access(t,p)=v)
        variant 2+u-l }
      m := (mean !l !u);
      assert { l <= m and m <= u };
      if t[!m] < v then
        l := !m + 1
      else if t[!m] > v then
        u := !m - 1
      else begin
        p := !m; (* break => *) l := !u + 1
      end       
    done
  end
  { (1 <= p and p <= N and access(t,p)=v) or (p = 0 and not In(t,1,N)) }

